### PR TITLE
Add mixer tests for the Istio authn filter

### PR DESCRIPTION
**Pull request by lei-tang**
Add the following Mixer tests for Istio authn filter
- Test when requiring JWT for peer and binding to peer, the authn attributes in the actual check and report calls match those in the expected check call
- Test when requiring JWT for peer and binding to origin, but no method specified in origin policy, the request will be rejected by Istio authn filter.
- Test when when requiring JWT for origin and binding to origin, the authn attributes in the actual check and report calls match those in the expected check call.
- Test when requiring JWT for origin and no binding to origin, the authn attributes in the actual check and report calls match those in the expected check call.
- Test when the HTTP request is rejected by the Istio authn filter for peer JWT authentication, the response code and the response message is as expected.
- Test when the HTTP request is rejected by the Istio authn filter for origin JWT authentication, the response code and the response message is as expected.
- Test when the Istio authn filter requires mTLS for peer connection, the non mTLS connection is rejected and the response code and the response message are as expected.
- Test when the Istio authn filter requires TLS for peer connection, the non TLS connection is rejected and the response code and the response message are as expected.

### CODE REVIEW

File: mixer/test/client/istio_authn/authn_report_test.go [hunk 1]
```diff
@@ -0,0 +1,323 @@
+// Copyright 2018 Istio Authors. All Rights Reserved.
```
Comment (diemtvu) [inline mixer/test/client/istio_authn/authn_report_test.go]: Is this report only, or for both report and check. We may need want for check, as authZ will need that.

**Comment by lei-tang** [comment]
This PR is for #5077

File: mixer/test/client/istio_authn/authn_report_test.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+  "response.size": 0,
+  "response.duration": "*",
+  "response.code": 200,
+  "response.headers": {
+     "date": "*",
+     "content-length": "0",
+     ":status": "200",
+     "server": "envoy"
+  },
+  "request.auth.audiences": "aud1",
```
Comment (diemtvu) [inline mixer/test/client/istio_authn/authn_report_test.go]: no request.auth.principal?

File: mixer/test/client/istio_authn/authn_report_test.go [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+                "name": "istio_authn",
+                "config": {
+                  "policy": {
+                    "origins": [
+                      {
+                        "jwt": {
+                          "issuer": "issuer@foo.com",
+                          "jwks_uri": "http://localhost:8081/"
+                        }
+                      }
```
Comment (diemtvu) [inline mixer/test/client/istio_authn/authn_report_test.go]: Need another test for with binding to origin. Also, can we have another for peer / mTLS case?

File: mixer/test/client/istio_authn/authn_report_test.go (hunk 1 above)
Comment (lei-tang) [reply mixer/test/client/istio_authn/authn_report_test.go]: Added the tests for check.

File: mixer/test/client/istio_authn/authn_report_test.go (hunk 2 above)
Comment (lei-tang) [reply mixer/test/client/istio_authn/authn_report_test.go]: Added the test for request.auth.principal in TestAuthnCheckReportAttributesBoundToOrigin().

File: mixer/test/client/istio_authn/authn_report_test.go (hunk 3 above)
Comment (lei-tang) [reply mixer/test/client/istio_authn/authn_report_test.go]: Added the tests with binding to origin. For peer/mTLS tests, I prefer to add them in separate PRs.

File: mixer/test/client/istio_authn_bound_origin/authn_report_test.go [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+     "sub": "sub@foo.com",
+     "aud": "aud1",
+     "some-other-string-claims": "some-claims-kept"
+  }
+}
+`
+
+// TODO: convert to v2, real clients use bootstrap v2 and all configs are switching !!!
+// The envoy config template with Istio authn filter that binds to the origin
+const envoyConfTempl = `
```
[the remark this answers is not in the document]
Comment (lei-tang) [reply mixer/test/client/istio_authn_bound_origin/authn_report_test.go]: Done: add {{.IstioAuthnFilter}} as an input to envoyConfTempl in envoy_conf.go.

File: mixer/test/client/env/envoy_conf.go [hunk 5]
```diff
@@ -101,6 +102,7 @@ const envoyConfTempl = `
             ],
             "filters": [
 {{.FaultFilter}}
+{{.IstioAuthnFilter}}
```
Comment (qiwzhang) [inline mixer/test/client/env/envoy_conf.go]: How about just use {{.FaultFilter}},  but rename it to {{.OtherFilters}}

File: mixer/test/client/env/setup.go [hunk 6]
```diff
@@ -78,6 +80,20 @@ func NewTestSetup(name uint16, t *testing.T) *TestSetup {
 	}
 }
 
+// NewTestSetupWithEnvoyConfig creates a new test setup with an Envoy config
+// "name" has to be defined in ports.go
+func NewTestSetupWithEnvoyConfig(name uint16, config string, t *testing.T) *TestSetup {
```
Comment (qiwzhang) [inline mixer/test/client/env/setup.go]: Do we need this function?

File: mixer/test/client/istio_authn_no_bound_origin/authn_report_test.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+	// Issues a GET echo request with 0 size body
+	tag := "OKGet"
+
+	// Add jwt_auth header to be consumed by Istio authn filter
+	headers := map[string]string{}
+	headers[secIstioAuthUserInfoHeaderKey] =
+		base64.StdEncoding.EncodeToString([]byte(secIstioAuthUserinfoHeaderValue))
+
+	if _, _, err := env.HTTPGetWithHeaders(url, headers); err != nil {
```
Comment (qiwzhang) [inline mixer/test/client/istio_authn_no_bound_origin/authn_report_test.go]: please add a test to test rejected requests by authn

Comment (lei-tang) [reply mixer/test/client/istio_authn_no_bound_origin/authn_report_test.go]: Done: added two tests to test rejected requests by authn.

File: mixer/test/client/istio_authn_origin_reject_no_jwt/authn_origin_reject_no_jwt_test.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	tag := "OKGet"
+
+	// No jwt_auth header to be consumed by Istio authn filter.
+	// The request will be rejected by Istio authn filter.
+	code, _, err := env.HTTPGet(url)
+	if err != nil {
+		t.Errorf("Failed in request %s: %v", tag, err)
+	}
+	// Verify that the http request is rejected
+	if code != 401 {
```
Comment (qiwzhang) [inline mixer/test/client/istio_authn_origin_reject_no_jwt/authn_origin_reject_no_jwt_test.go]: do we have error message in the response body?  if we do, could we verify it too?

File: mixer/test/client/env/setup.go [hunk 9]
```diff
@@ -78,6 +79,18 @@ func NewTestSetup(name uint16, t *testing.T) *TestSetup {
 	}
 }
 
+// NewTestSetupWithAuthnConfig creates a new test setup with an Istio authn config
+// "name" has to be defined in ports.go
+func NewTestSetupWithAuthnConfig(name uint16, config string, t *testing.T) *TestSetup {
```
Comment (qiwzhang) [inline mixer/test/client/env/setup.go]: can we remove this function?

File: mixer/test/client/env/setup.go [hunk 10]
```diff
@@ -34,14 +34,15 @@ type TestSetup struct {
...
-	stress        bool
-	faultInject   bool
-	noMixer       bool
-	mfConfVersion string
+	envoy    *Envoy
+	mixer    *MixerServer
+	backend  *HTTPServer
+	testName uint16
+	stress   bool
+	//The configurations of the filters before the Mixer filter
```
Comment (qiwzhang) [inline mixer/test/client/env/setup.go]: remove the comment so all fields can be nicely aligned.

File: mixer/test/client/istio_authn_origin_reject_no_jwt/authn_origin_reject_no_jwt_test.go (hunk 8 above)
Comment (lei-tang) [reply mixer/test/client/istio_authn_origin_reject_no_jwt/authn_origin_reject_no_jwt_test.go]: Done: added the verification of the error messages.

File: mixer/test/client/env/setup.go (hunk 9 above)
Comment (lei-tang) [reply mixer/test/client/env/setup.go]: Done.

File: mixer/test/client/env/setup.go (hunk 10 above)
Comment (lei-tang) [reply mixer/test/client/env/setup.go]: Done.

File: mixer/test/client/istio_authn/authn_report_test.go (hunk 3 above)
Comment (lei-tang) [reply mixer/test/client/istio_authn/authn_report_test.go]: Added peer, TLS, and mTLS tests. The following is the tests added in this PR: 
- Test when requiring JWT for peer and binding to peer, the authn attributes in the actual check and report calls match those in the expected check call
- Test when requiring JWT for peer and binding to origin, but no method specified in origin policy, the request will be rejected by Istio authn filter.
- Test when when requiring JWT for origin and binding to origin, the authn attributes in the actual check and report calls match those in the expected check call.
- Test when requiring JWT for origin and no binding to origin, the authn attributes in the actual check and report calls match those in the expected check call.
- Test when the HTTP request is rejected by the Istio authn filter for peer JWT authentication, the response code and the response message is as expected.
- Test when the HTTP request is rejected by the Istio authn filter for origin JWT authentication, the response code and the response message is as expected.
- Test when the Istio authn filter requires mTLS for peer connection, the non mTLS connection is rejected and the response code and the response message are as expected.
- Test when the Istio authn filter requires TLS for peer connection, the non TLS connection is rejected and the response code and the response message are as expected.

**Comment by lei-tang** [comment]
Created the PR https://github.com/istio/istio/pull/5267 on the release-0.8 branch to add the tests in this PR to the release-0.8 branch.

**Comment by lei-tang** [comment]
#5267

**Comment by lei-tang** [comment]
This PR has been merged into release 0.8 branch with https://github.com/istio/istio/pull/5267.

**Pull request closed without merging**